core/modules/block/block.module

```diff
@@ -279,6 +279,19 @@ function block_custom_block_save(array $edit, $delta = NULL) {
   $config->setData($block);
   $config->save();
 
+  // Save each file as permanent, preventing it from being deleted.
+  $fids = filter_parse_file_fids($block['body']['value']);
+  $files = file_load_multiple($fids);
+  foreach ($files as $fid => $file) {
+    if ($file && $file->status !== FILE_STATUS_PERMANENT) {
+      $file->status = FILE_STATUS_PERMANENT;
+      $file->save();
+    }
+    // Essentially this claims that the file is "self-referencing", which means
+    // it will never be deleted until a file management tool is available.
+    file_usage_add($file, 'file', 'file', $file->fid);
+  }
+
```

Comment:
> These blocks can't really be used except in a layout, which means on layout save, this process will happen twice.

I thought this would be the case, but it turns out that this is not the case. Once a block has been "converted" into reusable block (from block.module) it becomes a `BlockLegacy` object instead of a `BlockText` object. So this code in `block_custom_block_save()` is only triggered for reusable blocks, and the code in `Layout:save()` only applies to non-reusable (`BlockText`) blocks.

This code can also be simplified because `file_usage_add()` also marks files as permanent... so we might not really need to centralize this at all (but we should add some inline documentation).